webui/react/src/components/TableActionBar.tsx

```diff
@@ -148,54 +148,93 @@ const TableActionBar: React.FC<Props> = ({
-    return experiments.filter(Loadable.isLoaded).reduce(
-      (acc, experiment) => {
-        acc[experiment.data.experiment.id] = getProjectExperimentForExperimentItem(
-          experiment.data.experiment,
-          project,
-        );
+  const selectedProjectExperimentMap = useMemo(() => {
+    return [...selectedExperimentsMap].reduce(
+      (acc, [id, { experiment }]) => {
+        acc[id] = getProjectExperimentForExperimentItem(experiment, project);
         return acc;
       },
       {} as Record<number, ProjectExperiment>,
     );
-  }, [experiments, project]);
+  }, [selectedExperimentsMap, project]);
 
   const selectedExperiments = useMemo(
-    () =>
-      Array.from(selectedExperimentIds.keys()).flatMap((id) =>
-        id in experimentMap ? [experimentMap[id]] : [],
-      ),
-    [experimentMap, selectedExperimentIds],
+    () => Object.values(selectedProjectExperimentMap),
+    [selectedProjectExperimentMap],
   );
 
   const availableBatchActions = useMemo(() => {
-    if (selectAll)
+    if (selection.type === 'ALL_EXCEPT')
       return batchActions.filter((action) => action !== ExperimentAction.OpenTensorBoard);
-    const experiments = experimentIds.map((id) => experimentMap[id]) ?? [];
-    return getActionsForExperimentsUnion(experiments, [...batchActions], permissions);
+    return getActionsForExperimentsUnion(selectedExperiments, [...batchActions], permissions);
     // Spreading batchActions is so TypeScript doesn't complain that it's readonly.
-  }, [experimentIds, experimentMap, permissions, selectAll]);
+  }, [permissions, selectedExperiments, selection.type]);
+
+  const completeFilterSet: Loadable<FilterFormSetWithoutId | undefined> = useMemo(() => {
+    if (selection.type === 'ONLY_IN') {
+      return NotLoaded;
+    }
+    return validFilterSet.map((fs) => {
+      // filter group consisting of all excluded ids and project id
+      const filterGroup = {
+        children: [
+          ...selection.exclusions.map((e) => ({
+            columnName: 'id',
+            kind: FormKind.Field,
+            location: V1LocationType.EXPERIMENT,
+            operator: Operator.NotEq,
+            type: V1ColumnType.NUMBER,
+            value: e,
+          })),
+          {
+            columnName: 'projectId',
+            kind: FormKind.Field,
+            location: V1LocationType.EXPERIMENT,
+            operator: Operator.Eq,
+            type: V1ColumnType.NUMBER,
+            value: project.id,
+          },
+        ],
+        conjunction: Conjunction.And,
+        kind: FormKind.Group,
+      };
+      return {
+        ...fs,
+        filterGroup: {
+          children: [filterGroup, fs.filterGroup],
+          conjunction: Conjunction.And,
+          kind: FormKind.Group,
+        },
+      };
+    });
+  }, [selection, validFilterSet, project.id]);
 
   const sendBatchActions = useCallback(
     async (action: BatchAction): Promise<BulkActionResult | void> => {
+      let params: BulkActionParams;
```

Comment:
idk if you like this way, but i think we want `params` to be const. if not preferred, ignore this

```ts
      const params: BulkActionParams = (() => {
        if (selection.type === 'ONLY_IN') {
          // TODO: when the selection spans pages, this check will not work --
          // should probably warn the user and/handle on the backend
          return {
            experimentIds: selectedExperiments.filter((exp) => !exp.unmanaged).map((exp) => exp.id),
          };
        } else {
          return completeFilterSet.match({
            _: () => ({ experimentIds: [] }),
            Loaded: (filterSet) => {
              return { experimentIds: [], searchFilter: JSON.stringify(filterSet) };
            },
          });
        }
      })();
```

Reply:
i'll use a ternary here -- an iife is overkill here because we're not introducing any other variables.